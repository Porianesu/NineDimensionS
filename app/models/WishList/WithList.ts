import { destroy, getParent, Instance, SnapshotIn, SnapshotOut, types } from "mobx-state-tree"

export const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: types.optional(types.string, ""),
  })
  .actions((self) => ({
    changeName(newName) {
      self.name = newName
    },
    changePrice(newPrice) {
      self.price = newPrice
    },
    changeImage(newImage) {
      self.image = newImage
    },
    remove() {
      getParent<WishListType>(self, 2).remove(self as unknown as WishListItemType)
    },
  }))
export interface WishListItemType extends Instance<typeof WishListItem> {}
export interface WishListItemSnapshotOut extends SnapshotOut<typeof WishListItem> {}
export interface WishListItemSnapshotIn extends SnapshotIn<typeof WishListItem> {}
export const WishList = types
  .model({
    priceTimer: types.maybeNull(types.number),
    items: types.optional(types.array(WishListItem), []),
  })
  .actions((self) => ({
    add: (item) => {
      self.items.push(item)
    },
    startPriceChange: () => {
      if (self.priceTimer) {
        clearInterval(self.priceTimer)
        self.priceTimer = null
      } else {
        self.priceTimer = setInterval(() => {
          self.items[0].changePrice(self.items[0].price + 1)
        }, 1000) as unknown as number
      }
    },
    remove: (targetItem: WishListItemType) => {
      destroy(targetItem)
      // self.items.splice(self.items.indexOf(targetItem), 1)
    },
  }))
  .actions((self) => ({
    initWishList: async () => {
      self.items.replace([
        WishListItem.create({
          name: "测试1",
          price: 123,
          image: "",
        }),
      ])
    },
  }))
  .views((self) => {
    return {
      get totalPrice() {
        return self.items.reduce<number>((sum, entry) => {
          return sum + entry.price
        }, 0)
      },
    }
  })
export interface WishListType extends Instance<typeof WishList> {}
export interface WishListSnapshotOut extends SnapshotOut<typeof WishList> {}
export interface WishListSnapshotIn extends SnapshotIn<typeof WishList> {}
