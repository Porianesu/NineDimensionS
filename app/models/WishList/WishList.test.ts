import { WishList, WishListItem } from "./WithList"
import { getSnapshot, onPatch } from "mobx-state-tree"
import { reaction } from "mobx"

it("can create an instance of a model ", () => {
  const item = WishListItem.create({
    name: "测试1",
    price: 12.5,
  })
  expect(item.price).toBe(12.5)
  expect(item.image).toBe("")
  expect(item.name).toBe("测试1")
  item.changeName("测试2")
  expect(item.name).toBe("测试2")
})

it("can create a wishListView", function () {
  const list = WishList.create()
  const patches = []
  onPatch(list, (patch) => {
    patches.push(patch)
  })
  list.add({ name: "测试3", price: 11, image: "" })
  expect(getSnapshot(list)).toMatchSnapshot()
  list.items[0].changeName("测试4")
})
it("can calculate the total price of a wishlist ", function () {
  const list = WishList.create()
  let changed = 0
  reaction(
    () => list.totalPrice,
    () => changed++,
  )
  expect(list.totalPrice).toBe(0)
  expect(changed).toBe(0)
  list.add(WishListItem.create({ name: "测试3", price: 11, image: "" }))
  expect(list.totalPrice).toBe(11)
  expect(changed).toBe(1)
  list.items[0].changePrice(12)
  expect(changed).toBe(2)
  list.add(WishListItem.create({ name: "测试4", price: 55, image: "" }))
})
