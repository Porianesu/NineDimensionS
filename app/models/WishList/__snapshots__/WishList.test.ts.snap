// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a wishList 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "测试3",
      "price": 11,
    },
  ],
}
`;
